File: products/compliance/src/reporters/badge.ts
import type { ScanResult, Framework } from '../types.js';

/**
 * Calculate the compliance score as a percentage.
 *
 * Score = (totalRules - findings) / totalRules * 100
 * Clamped to [0, 100].
 */
export function calculateScore(totalFindings: number, totalRules: number): number {
  if (totalRules <= 0) return 0;
  const score = ((totalRules - totalFindings) / totalRules) * 100;
  return Math.max(0, Math.min(100, Math.round(score * 10) / 10));
}

/**
 * Return the badge color based on score thresholds.
 *
 * - Green (#4c1) for scores above 80%
 * - Yellow (#dfb317) for scores between 60% and 80% (inclusive)
 * - Red (#e05d44) for scores below 60%
 */
export function scoreColor(score: number): string {
  if (score > 80) return '#4c1';
  if (score >= 60) return '#dfb317';
  return '#e05d44';
}

/**
 * Build a human-readable label from the scanned frameworks.
 */
function frameworkLabel(frameworks: Framework[]): string {
  if (frameworks.length === 0) return 'compliance';
  return frameworks.map((f) => f.toUpperCase()).join(' + ');
}

/**
 * Generate an SVG compliance badge in shields.io flat style.
 *
 * The badge displays the framework name(s) on the left half and the
 * compliance score percentage on the right half, colored according to
 * the score thresholds.
 */
export function generateBadge(result: ScanResult, totalRules: number): string {
  const score = calculateScore(result.summary.total, totalRules);
  const color = scoreColor(score);
  const label = frameworkLabel(result.metadata.frameworks);
  const scoreText = `${score}%`;

  // Calculate text widths (approximate: 6.5px per character + padding)
  const labelWidth = Math.max(label.length * 6.5 + 10, 40);
  const valueWidth = Math.max(scoreText.length * 6.5 + 10, 40);
  const totalWidth = labelWidth + valueWidth;

  return `<svg xmlns="http://www.w3.org/2000/svg" width="${totalWidth}" height="20" role="img" aria-label="${label}: ${scoreText}">
  <title>${label}: ${scoreText}</title>
  <linearGradient id="s" x2="0" y2="100%">
    <stop offset="0" stop-color="#bbb" stop-opacity=".1"/>
    <stop offset="1" stop-opacity=".1"/>
  </linearGradient>
  <clipPath id="r">
    <rect width="${totalWidth}" height="20" rx="3" fill="#fff"/>
  </clipPath>
  <g clip-path="url(#r)">
    <rect width="${labelWidth}" height="20" fill="#555"/>
    <rect x="${labelWidth}" width="${valueWidth}" height="20" fill="${color}"/>
    <rect width="${totalWidth}" height="20" fill="url(#s)"/>
  </g>
  <g fill="#fff" text-anchor="middle" font-family="Verdana,Geneva,DejaVu Sans,sans-serif" text-rendering="geometricPrecision" font-size="11">
    <text x="${labelWidth / 2}" y="14">${label}</text>
    <text x="${labelWidth + valueWidth / 2}" y="14">${scoreText}</text>
  </g>
</svg>`;
}

File: internal/quality/compliance/reporters/badge.go
package reporters

import (
	"fmt"

	"github.com/rishav1305/soul/internal/quality/compliance/scan"
)

// CalculateScore computes a compliance score from 0-100 based on findings.
// Formula: 100 - (critical*10 + high*5 + medium*2 + low*1), clamped to [0,100].
func CalculateScore(result *scan.ScanResult) int {
	s := result.Summary.BySeverity
	score := 100 - (s["critical"]*10 + s["high"]*5 + s["medium"]*2 + s["low"]*1)
	if score < 0 {
		score = 0
	}
	if score > 100 {
		score = 100
	}
	return score
}

// GenerateBadge returns an SVG badge showing the compliance score.
func GenerateBadge(result *scan.ScanResult) string {
	score := CalculateScore(result)

	var color string
	switch {
	case score >= 90:
		color = "#4c1"    // green
	case score >= 70:
		color = "#dfb317" // yellow
	case score >= 50:
		color = "#fe7d37" // orange
	default:
		color = "#e05d44" // red
	}

	return fmt.Sprintf(`<svg xmlns="http://www.w3.org/2000/svg" width="150" height="20">
  <linearGradient id="b" x2="0" y2="100%%">
    <stop offset="0" stop-color="#bbb" stop-opacity=".1"/>
    <stop offset="1" stop-opacity=".1"/>
  </linearGradient>
  <mask id="a">
    <rect width="150" height="20" rx="3" fill="#fff"/>
  </mask>
  <g mask="url(#a)">
    <rect width="80" height="20" fill="#555"/>
    <rect x="80" width="70" height="20" fill="%s"/>
    <rect width="150" height="20" fill="url(#b)"/>
  </g>
  <g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11">
    <text x="40" y="15" fill="#010101" fill-opacity=".3">compliance</text>
    <text x="40" y="14">compliance</text>
    <text x="115" y="15" fill="#010101" fill-opacity=".3">%d%%</text>
    <text x="115" y="14">%d%%</text>
  </g>
</svg>`, color, score, score)
}
